### PR TITLE
fix: prevent observing inconsistent tracker states

### PR DESCRIPTION
**Pull request by tustvold**
As was correctly pointed out by @domodwyer the current API allows observing an inconsistent "snapshot" of the state of a tracker.

This PR alters the tracker to return a TrackerStatus that prevents observing "inconsistent" states, in particular a larger pending_count than total_count, and makes it clear what is guaranteed to be "converged" at what point.

### CODE REVIEW

File: server/src/tracker.rs [hunk 1]
```diff
@@ -108,6 +108,41 @@ struct TrackerState {
...
 
+/// The status of the tracker
+#[derive(Debug, Clone)]
+pub enum TrackerStatus {
+    /// More futures can be registered
+    Creating,
+
+    /// No more futures can be registered
+    ///
+    /// `pending_count` and `cpu_nanos` are best-effort -
```
Comment (tustvold) [inline server/src/tracker.rs:119]: I'm not sure what the alternative would be, as necessarily a future could finish at the same time an API request is being constructed, and cpu_nanos is only updated on exiting from poll, but I thought it worth pointing out.

Comment (alamb) [reply server/src/tracker.rs:119]: I think it is a good idea to be explicit here about being best effort -- I think it makes the code clearer for a larger audience.

**Review by alamb — APPROVED** [review]
I went through this PR and it makes sense to me. I like the change to explicitly enumerate state and what fields / values are valid for each state. 👍 

Nice work

**Comment by alamb** [comment]
@tustvold do you plan to merge this one? (I ask because a see a few other prs messing with the registry too and I wonder if they might conflict with each other)

**Review by domodwyer — APPROVED** [review]
Looks awesome, thanks!

A key point that eluded me at first is this synchronisation makes use of rust's aliasing guarantees - specifically around the clone-and-then-drop of a `TrackerRegistration`. It's impossible to clone in one thread, and drop in another thread (which would cause a potentially incorrect read of the ref count I believe?) because you can only give the cloned handle to another thread through a synchronised mechanism in safe rust, therefore ensuring it has seen the +1 from the clone. Neat! A comment explaining it would be cool if you fancy.

I hadn't thought of Rust's ownership model allowing to optimise this sort of thing - it's cool!

**Comment by alamb** [comment]
Note this is part of #942

**Comment by tustvold** [comment]
@alamb Was waiting to give @domodwyer a chance to review

>  It's impossible to clone in one thread, and drop in another thread (which would cause a potentially incorrect read of the ref count I believe?)

If I follow correctly, the fear is that the only guarantee provided by atomic operations is that writes to the same memory address are guaranteed to be serializable, but on their own there is no guarantee as to what their order is.

Within a single thread, it would be invalid, irrespective of atomics, for the decrement to be ordered prior to the increment. However, if drop occurred in a different thread the decrement could commit before the increment.

As the increment and decrement are coupled into the constructor and destructor of the type respectively, the only way for this to occur would be if the data were sent from one to another. Any sane message passing system will ensure writes prior to the send are visible in the destination thread, and so this situation cannot occur.

I'm not sure I understand your comment about Rust's aliasing guarantees, but I think the thing that makes this safe is the nature of RAII - if that's what you're referring to then I agree :+1:. FWIW this isn't unique to Rust, std::shared_ptr in C++ makes the same optimisation.

The comment at the top of the tracker module has the following, let me know if there is a clearer way to state this

```
//! Note: this proof ignores the complexity of moving Trackers, TrackedFutures,
//! etc... between threads as any such functionality must perform the necessary
//! synchronisation to be well-formed.
```

**Pull request merged**